src/Hamiltonians/Sum.jl

```diff
@@ -0,0 +1,142 @@
+end
+
+function VectorInterface.add(h1::AbstractHamiltonian, h2::AbstractHamiltonian, a::Number, b::Number; weight=0.5)
+    return HamiltonianSum(a*h1, b*h2; weight)
+end
+
+starting_address(s::HamiltonianSum) = starting_address(s.h1)
+
+function allows_address_type(s::HamiltonianSum, ::Type{A}) where {A}
+    return allows_address_type(s.h2, A) && allows_address_type(s.h1, A)
+end
+
+function LOStructure(::Type{<:HamiltonianSum{<:Any,H1,H2}}) where {H1,H2}
+    l1 = LOStructure(H1)
+    l2 = LOStructure(H2)
+    if l1 == IsDiagonal() && l2 == IsDiagonal()
+        return IsDiagonal()
+    elseif l1 == IsHermitian() && l2 == IsHermitian()
+        return IsHermitian()
+    elseif l1 != AdjointUnknown() && l2 != AdjointUnknown()
+        return AdjointKnown()
+    else
+        return AdjointUnknown()
+    end
+end
+
+function LinearAlgebra.adjoint(s::HamiltonianSum)
+    return HamiltonianSum(s.h1', s.h2'; weight=s.weight)
+end
+
+function has_iterable_offdiagonals(::Type{<:HamiltonianSum{<:Any,H1,H2}}) where {H1,H2}
+    return has_iterable_offdiagonals(H1) && has_iterable_offdiagonals(H2)
+end
+
+function has_random_offdiagonal(::Type{<:HamiltonianSum{<:Any,H1,H2}}) where {H1, H2}
+    return has_random_offdiagonal(H1) && has_random_offdiagonal(H2)
+end
+
+struct SumColumn{A,T,O<:HamiltonianSum{T},C1,C2} <: AbstractOperatorColumn{A,T,O}
+    operator::O
+    address::A
+    col1::C1
+    col2::C2
+    weight::Float64
+end
+function operator_column(s::HamiltonianSum, add)
+    return SumColumn(s, add, operator_column(s.h1,add), operator_column(s.h2,add), s.weight)
+end
+
+parent_operator(c::SumColumn) = c.operator
+starting_address(c::SumColumn) = c.address
+num_offdiagonals(c::SumColumn) = num_offdiagonals(c.col1) + num_offdiagonals(c.col2)
+
+function diagonal_element(c::SumColumn{<:Any,T}) where {T}
+    return T(diagonal_element(c.col1) + diagonal_element(c.col2))
+end
+
+function random_offdiagonal(c::SumColumn{<:Any,T}) where {T}
+    if rand() < c.weight
+        add, prob, val = random_offdiagonal(c.col1)
+        return add, prob*c.weight, T(val)
+    else
+        add, prob, val = random_offdiagonal(c.col2)
+        return add, prob*(1-c.weight), T(val)
+    end
+end
+
+struct SumOffdiagonals{A,T,O<:HamiltonianSum{T},OD1,OD2}
+    operator::O
+    address::A
+    ods1::OD1
+    ods2::OD2
+end
+function offdiagonals(c::SumColumn)
+    return SumOffdiagonals(c.operator, c.address, offdiagonals(c.col1), offdiagonals(c.col2))
+end
+
+Base.IteratorSize(::SumOffdiagonals) = Base.SizeUnknown()
+Base.eltype(::SumOffdiagonals{A,T}) where {A,T} = Pair{A,T}
+
+function Base.iterate(o::SumOffdiagonals)
+    first = iterate(o.ods1)
+    if isnothing(first)
+        first = iterate(o.ods2)
+        if isnothing(first)
+            return nothing
+        end
+        (add, val), state = first
+        return add => val, (state, false)
+    end
+    (add, val), state = first
+    return add => val, (state, true)
```

Comment:
This is all good, but I'm a bit confused why the variable is named `first`. Would the following make sense?
```suggestion
    if isnothing(first) # continue with second Hamiltonian
        second = iterate(o.ods2)
        if isnothing(second)
            return nothing
        end
        (add, val), state = second
        return add => val, (state, false)
    else
        (add, val), state = first
        return add => val, (state, true)
    end
```

Reply:
Yes the names were unclear, I've renamed them now to agree with `HamiltonianProduct`.